File: orchestrator/agents/release_agent.py
"""Release-readiness agent: the last node in every scenario graph. By the
time it runs, the entry-gate policy `release_requires_clean_policy_history`
has already refused to let it start if any policy violation is outstanding,
and its own `requires_approval=True` flag means it only executes at all
after the approval checkpoint has granted access -- so reaching this
function's body is itself evidence the run is clean. It compiles that
evidence into a release note.
"""
import os
from datetime import datetime, timezone


def prepare_release(ctx, node, project_root: str, release_notes_rel_path: str) -> dict:
    lines = [
        f"# Release note -- {ctx.scenario} (run {ctx.run_id})",
        "",
        f"Generated {datetime.now(timezone.utc).isoformat()}",
        "",
        f"**Original request:** {ctx.raw_request}",
        "",
        "## Decision lineage",
        "",
    ]
    for d in ctx.decisions:
        lines.append(f"- **[{d.stage}/{d.node_id}]** {d.summary}")
        lines.append(f"  - rationale: {d.rationale}")
        for alt in d.alternatives_considered:
            lines.append(f"  - rejected alternative: {alt}")
    lines += ["", "## Artifacts touched", ""]
    for a in ctx.artifacts:
        lines.append(f"- {a}")
    lines += ["", "## Approvals", ""]
    for a in ctx.approvals:
        lines.append(f"- node `{a['node_id']}`: {'APPROVED' if a['approved'] else 'WITHHELD'} by {a['approver']} -- {a['rationale']}")
    lines += ["", "## Policy violations recorded during this run", ""]
    if ctx.policy_violations:
        for v in ctx.policy_violations:
            lines.append(f"- `{v['node_id']}` / {v['policy']}: {v['message']}")
    else:
        lines.append("- none")
    content = "\n".join(lines) + "\n"

    abs_path = os.path.join(project_root, release_notes_rel_path)
    os.makedirs(os.path.dirname(abs_path), exist_ok=True)
    with open(abs_path, "w", encoding="utf-8") as f:
        f.write(content)
    ctx.add_artifact(release_notes_rel_path)

    ctx.record_decision(
        stage=node.stage,
        node_id=node.id,
        summary="Marked release-ready and wrote release note.",
        rationale="Reached this node only because the entry-gate policy found zero unresolved policy "
        "violations and the approval checkpoint had already granted access.",
    )
    return {"release_ready": True, "release_note_path": release_notes_rel_path}

File: src/main/java/com/example/urlshortener/orchestrator/agents/ReleaseAgent.java
package com.example.urlshortener.orchestrator.agents;

import com.example.urlshortener.orchestrator.Decision;
import com.example.urlshortener.orchestrator.Node;
import com.example.urlshortener.orchestrator.SharedContext;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Release-readiness agent: the last node in every scenario graph. By the
 * time it runs, the entry-gate release policy has already refused to let it
 * start if any policy violation is outstanding, and its own
 * requiresApproval=true flag means it only executes at all after the
 * approval checkpoint has granted access -- so reaching this function's body
 * is itself evidence the run is clean. Direct port of
 * orchestrator/agents/release_agent.py.
 */
public final class ReleaseAgent {
    private ReleaseAgent() {}

    public static Map<String, Object> prepareRelease(SharedContext ctx, Node node, Path projectRoot, String releaseNotesRelPath) {
        StringBuilder sb = new StringBuilder();
        sb.append("# Release note -- ").append(ctx.scenario).append(" (run ").append(ctx.runId).append(")\n\n");
        sb.append("Generated ").append(Instant.now()).append("\n\n");
        sb.append("**Original request:** ").append(ctx.rawRequest).append("\n\n");
        sb.append("## Decision lineage\n\n");
        for (Decision d : ctx.decisions) {
            sb.append("- **[").append(d.stage()).append("/").append(d.nodeId()).append("]** ").append(d.summary()).append("\n");
            sb.append("  - rationale: ").append(d.rationale()).append("\n");
            for (String alt : d.alternativesConsidered()) sb.append("  - rejected alternative: ").append(alt).append("\n");
        }
        sb.append("\n## Artifacts touched\n\n");
        for (String a : ctx.artifacts) sb.append("- ").append(a).append("\n");
        sb.append("\n## Approvals\n\n");
        for (Map<String, Object> a : ctx.approvals) {
            sb.append("- node `").append(a.get("node_id")).append("`: ")
                    .append(Boolean.TRUE.equals(a.get("approved")) ? "APPROVED" : "WITHHELD")
                    .append(" by ").append(a.get("approver")).append(" -- ").append(a.get("rationale")).append("\n");
        }
        sb.append("\n## Policy violations recorded during this run\n\n");
        if (ctx.policyViolations.isEmpty()) {
            sb.append("- none\n");
        } else {
            for (Map<String, Object> v : ctx.policyViolations) {
                sb.append("- `").append(v.get("node_id")).append("` / ").append(v.get("policy")).append(": ").append(v.get("message")).append("\n");
            }
        }

        Path abs = projectRoot.resolve(releaseNotesRelPath);
        try {
            Files.createDirectories(abs.getParent());
            Files.writeString(abs, sb.toString());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        ctx.addArtifact(releaseNotesRelPath);

        ctx.recordDecision(
                node.stage, node.id,
                "Marked release-ready and wrote release note.",
                "Reached this node only because the entry-gate policy found zero unresolved policy violations "
                        + "and the approval checkpoint had already granted access."
        );

        Map<String, Object> out = new LinkedHashMap<>();
        out.put("release_ready", true);
        out.put("release_note_path", releaseNotesRelPath);
        return out;
    }
}
